pkg/compact/compact.go

```diff
@@ -470,6 +483,182 @@ func (cg *Group) Resolution() int64 {
+type ProgressMetrics struct {
+	NumberOfCompactionRuns   *prometheus.GaugeVec
+	NumberOfCompactionBlocks *prometheus.GaugeVec
+}
+
+// ProgressCalculator updates the results/metrics from the compaction planning and downsampling simulations.
+type ProgressCalculator interface {
+	ProgressCalculate(ctx context.Context, groups []*Group) error
+}
+
+// CompactionSimulator contains a planner and ProgressMetrics, which are updated during the compaction simulation process
+type CompactionSimulator struct {
+	planner Planner
+	*ProgressMetrics
+}
+
+func NewCompactionSimulator(reg prometheus.Registerer, planner *tsdbBasedPlanner) *CompactionSimulator {
+	return &CompactionSimulator{
+		planner: planner,
+		ProgressMetrics: &ProgressMetrics{
+			NumberOfCompactionRuns: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+				Name: "thanos_compact_todo_compactions",
+				Help: "number of iterations to be done",
+			}, []string{"group"}),
+			NumberOfCompactionBlocks: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+				Name: "thanos_compact_todo_compaction_blocks",
+				Help: "number of blocks planned to be merged",
+			}, []string{"group"}),
+		},
+	}
+}
+
+func (ps *CompactionSimulator) ProgressCalculate(ctx context.Context, groups []*Group) error {
+	groupCompactions := make(map[string]int, len(groups))
+	groupBlocks := make(map[string]int, len(groups))
+
+	for len(groups) > 0 {
+		tmpGroups := make([]*Group, 0, len(groups))
+		for _, g := range groups {
+			if len(g.IDs()) == 1 {
+				continue
+			}
+			plan, err := ps.planner.Plan(ctx, g.metasByMinTime)
+			if err != nil {
+				return errors.Wrapf(err, "could not plan")
+			}
+			if len(plan) == 0 {
+				continue
+			}
+			groupCompactions[g.key]++
+
+			toRemove := make(map[ulid.ULID]struct{}, len(plan))
+			metas := make([]*tsdb.BlockMeta, 0, len(plan))
+			for _, p := range plan {
+				metas = append(metas, &p.BlockMeta)
+				toRemove[p.BlockMeta.ULID] = struct{}{}
+			}
+			g.deleteFromGroup(toRemove)
+
+			groupBlocks[g.key] += len(plan)
+
+			if len(g.metasByMinTime) == 0 {
+				continue
+			}
+
+			newMeta := tsdb.CompactBlockMetas(ulid.MustNew(uint64(time.Now().Unix()), nil), metas...)
+			if err := g.AppendMeta(&metadata.Meta{BlockMeta: *newMeta, Thanos: metadata.Thanos{Downsample: metadata.ThanosDownsample{Resolution: g.Resolution()}, Labels: g.Labels().Map()}}); err != nil {
+				return errors.Wrapf(err, "append meta")
+			}
+			tmpGroups = append(tmpGroups, g)
+		}
+
+		groups = tmpGroups
+	}
+
+	for key, iters := range groupCompactions {
+		ps.ProgressMetrics.NumberOfCompactionRuns.WithLabelValues(key).Add(float64(iters))
+		ps.ProgressMetrics.NumberOfCompactionBlocks.WithLabelValues(key).Add(float64(groupBlocks[key]))
+	}
+
+	return nil
+}
+
+// DownsampleMetrics contains Prometheus metrics related to downsampling progress.
+type DownsampleMetrics struct {
+	BlocksDownsampled *prometheus.GaugeVec
```

Comment:
Let's call it `NumberOfDownsampleBlocks` here? As we already have`NumberOfCompactionBlocks` in the compaction simulator metrics

Reply:
Yes, that makes it more consistent. Thanks!

Reply:
Thanks for the detailed review of the comments! Certainly learnt a lot about concise comments without going into implementation details. I've fixed the names and comments in a recent commit.